### ISSUE TITLE
Handling timed-count butterfly data in verification

### ISSUE THREAD

**Issue by johnvanbreda** [BiologicalRecordsCentre/iRecord#1274]
The iRecord Butterflies app has a single species timed count mode which generates a parent sample for each time it is used, containing a sub-sample and occurrence with a precise location every time an individual of the searched species is found.

Currently the individuals will be represented in the verification tool as separate rows in the data grid, requiring each to be verified separately (though bulk-verification tools can assist with this). A more efficient approach might be to represent rows based on an aggregation of parent sample/species, so each species in a timed count is only represented as a single row. Clicking on a row will show a map of all the individual data points, with summarised information (e.g. a list of included sex/stages, plus data collected at the sample level). Options will be available for the verifier to apply a decision to all the data in the sample, or to expand the sample to view the sub-sample and occurrence data and verify each individually.

Need to consider how this impacts on the filter tool. For example, if a verifier filters for only adults, then a sample containing a mixture of adults and juveniles may be returned. Summary data created by the aggregation query will only include info for the adults, but when they expand the sample we might opt to include just the adults, or have a warning that there are other records in the sample. Presumably we need to take care to ensure verification decisions don't apply to occurrences that don't match the filter.

If this is a change to the existing tool rather than a separate tool, consideration needs to be given to how casual records are displayed and handled - will the summary data match the existing details display for a record? 

Limitations of Elasticsearch aggregations also need to be considered, as they are less flexible for pagination than un-aggregated data.

**Comment by johnvanbreda** [comment]
@DavidRoy a few thoughts about this. The requirement as stated does need a fairly significant amount of work and really is a completely new verification tool, since all aspects of the existing tool would be affected. It would be quite specific to the way the butterfly timed counts are recorded and the way aggregated data tables work in Elasticsearch is a bit limiting.

An alternative thought would be to provide a more generic tool for "reviewing nearby records" as an add-on to the existing screen. When reviewing any dataset, there are often going to be situations where multiple records of one taxon are present in the vicinity. These might be on the same day, or might be from other visits but could still be verified at the same time. They could also be from the same recorder or otherwise - if you have ascertained that a plant is at a site, it might be easiest to be able to handle them all in one go. So the requirement might be:

* Add a new "nearby" tab to the record details tab.
* When a verifier opens this tab, a data table is shown listing the records, with date/abundance/sex/stage and record details links.
* The records are then shown as points on the map.
* Filter options available - "nearby" or "this visit" - nearby would be a distance based filter, this visit would be based on parent sample ID.
* Also a filter option "limit to this recorder" (on/off), default to on.
* Each record has a tickbox and there is a button to tick all. When the verification decision is applied, it will be applied to all the ticked records. 

This approach enhances the existing tool in a way that might be generically useful and should be a bit less work. The problems with filtering described above are also avoided.

Any thoughts? @kitenetter it would also be interesting to hear what you think.

**Comment by johnvanbreda** [comment]
A ballpark of 3 days.

**Comment by kitenetter** [comment]
I agree that this could work well. Downsides that I can envisage are:
- It's another layer of complexity for verifiers to take on board, but if confined to a separate tab then it should be easy enough for verifiers to ignore it if they prefer to do so
- Without trying it with some examples of the timed count data it's difficult to get a feel for how much of a time saver it would be. In order to use the new tool, I have first got to realise that there are multiple similar records, then I've got to click on a new tab, then I've got to select the records, then I've got to verify them. I'm not yet convinced that this is much easier than using the existing bulk verification options.
- Much will depend on how tightly the filters are defined for the new tab. Should it be for one species in one grid ref on one date? And for one recorder (or one user account)? Should we give people options for how to filter, or does that increase the complexity and result in increased errors due to to many records being dealt with at one?

Overall I think it comes down to whether a new tool is needed, or whether we should put more effort into supporting people to use the existing bulk tools. It would be helpful to seek feedback on this from some butterfly verifiers.

@johnvanbreda one way that we could improve the existing bulk verification would be to allow a right-click option for "filter by this value", i.e. if you are looking at the verification grid you could right click over a species name to "filter by this species" or right-click over a date to "filter by this date". Is that sort of functionality likely to be possible?

@DavidRoy this doesn't feel like such a high priority as some of the other things on out lists, but it depends on how much timed count data we have, and how fast it will increase. Do you have any info on that?

**Comment by johnvanbreda** [comment]
@kitenetter a right click to "filter by this value" is certainly possible. Can you not already do something similar by pasting a value into the filter box at the top of the column?

**Comment by kitenetter** [comment]
@johnvanbreda it is possible to copy and paste from the grid into the filter box, but it is quite fiddly to do. I reckon it takes about 8 mouse-clicks to highlight a full recorder name, paste it into the filter box and get it to take effect. If we had a right-click option to "filter by this value" that might reduce it to 2 clicks? 

But I realise this is a separate issue from the main one we are addressing here!

**Comment by johnvanbreda** [comment]
Agreed - I suggest another issue so it can be prioritised separately then.

**Comment by kitenetter** [comment]
Have added #1381 to address the filter-by-right-click suggestion.

@DavidRoy do you have any info on how quickly timed-count records are likely to accumulate, and how high a priority this is?

**Comment by DavidRoy** [comment]
@kitenetter high priority before next March.  It's holding up completion of work on the iRecord butterflies app surveys

**Comment by kitenetter** [comment]
Okay. @johnvanbreda I'm not sure what the dependencies are here - can we set up something to test, or do decisions have to be made about the way the app behaves before we can start implementing verification tools?

**Comment by kitenetter** [comment]
From discussion today, we agreed that we should implement an approach to this on the main main verification page (rather than setting up a custom page).

John will explore options for adding a “verify all records for this species in this sample” button to the verification box when verifying records that are part of a larger sample-subsample hierarchy.

Not clear to me whether we need to limit this to particular website/survey sources, or whether a generic approach to all sample-subsample hierarchies will work.

Limiting to a single sample-subsample set of records feels like a more practical way of implementing this (at least initially) as opposed to the more complex "records nearby" approach suggested [above](https://github.com/BiologicalRecordsCentre/iRecord/issues/1274#issuecomment-1177763529).

**Comment by DavidRoy** [comment]
I agree on the 'simple' approach to start with

**Comment by johnvanbreda** [comment]
Just to check the implementation before I commit this - I've added a "hierarchy" button next to the verification buttons, which only gets enabled if the selected record is part of a sample/sub-sample hierarchy (like a transect or timed count record):

![image](https://user-images.githubusercontent.com/2392594/219072758-b2d4d5b1-3fb0-4ed0-b9b2-24e233ba9740.png)

This button has the following hint if you hover over it:

> Only available for records that are part of a parent sample, such as a transect or timed count. Click to activate the button then 
 verification decisions will apply to all unverified records of the same taxon within the parent sample.

If the button is enabled and active, it gets depressed and coloured as follows:

![image](https://user-images.githubusercontent.com/2392594/219073834-270faee2-0f86-495b-a322-23bc81bcc538.png)

Then, for verification decisions, the popup shows a warning. This only works for verification (not query or redets) and only when a single record is selected, not in bulk mode. The warning looks like this:

![image](https://user-images.githubusercontent.com/2392594/219074603-c7a9fcf6-bcb0-4922-b4a7-3d5d1cb426d8.png)

Does that implementation sound OK @kitenetter @DavidRoy ?

**Comment by kitenetter** [comment]
I think that will work. 

How complex is it to add the query and/or redet functions? I can imagine that if someone gets a species wrong in one part of a timed count or transect they might get it wrong for multiple sub-sample records. But perhaps this is one of those things that needs some user-testing before we review whether there is a requirement for additional functionality.

For the hint associated with the button, can we add:

> Only available for records that are part of a parent sample, such as a transect or timed count**, on a single date**.

**Comment by johnvanbreda** [comment]
Coding for this now complete in the develop branch.

**Comment by kitenetter** [comment]
This is now live on the main site. I've tested it for a couple of bee-fly records that came from the UKBMS transects, and it seems to work! Two requests:
1. When the verification pop-up appears, is it possible for it to show how many records are being included, in the same way that happens for the pop-ups when applying bulk verifications? This would help reassure verifiers that the system is behaving the way they expect.
2. In the message that is included in the pop-up, can we add info on the date as well, so that it reads: "This verification decision will be applied to all records of the same taxon on the date and location of the parent sample (e.g. for all sections of a single transect on a single date)."

That is a little long-winded but I think it will help explain what is happening.

**Comment by johnvanbreda** [comment]
I've added info on the number of records in the parent sample (develop branch to be released at next full update). 

However I'm not sure about the second point - the limit is based on the parent sample ID, not on the date itself and although there will often be a 1:1 match between parent sample and date it won't always be the case. For example a recorder might walk a transect twice in one day, but the verification decision will only be applied to the records for the same walk as the current record, not both. The system design also allows for a parent sample to contain child samples on different dates (hypothetically a transect walk could be stopped overnight and continued the next day, or multi-day trapping could be recorded) in which case the parent sample would encompass records from different dates.

**Comment by johnvanbreda** [comment]
@kitenetter can this be closed given the comment above?

**Comment by kitenetter** [comment]
Yes, I missed that comment at the time - closing.
